### ISSUE TITLE
Remove -xnolibs argument when calling dtrace

### ISSUE THREAD

**Issue by nalimilan** [joyent/libuv#1478]
`Makefile.am` passes `-xnolibs` argument to `dtrace` at https://github.com/joyent/libuv/blob/master/Makefile.am#L321. This argument is no longer supported by dtrace >= 2.5. I'm not sure why it was present in the first place, I must admit I have no idea what happens in that line.

This has been spotted when packaging Julia (bundling libuv) for Fedora [1]. 

1: https://bugzilla.redhat.com/show_bug.cgi?id=1040517

**Comment by saghul** [comment]
I have no idea what it does... @tjfontaine can you help? Either way, you can disable dtrace probes when building Julia (there is a ./configure modifier for it), currently there are just 2 which are not of much use.

**Comment by nalimilan** [comment]
Currently I'm calling sed to remove this argument from the Makefile. Not a big deal, but the Fedora policy is to move fixes upstream so that everybody benefits from them, instead of slowly piling hacks. :-)

**Comment by saghul** [comment]
Sure! Hopefully someone can clarify why is that needed. I believe we mostly
build with dtrace support on OSX and FreeBSD.

**Comment by indutny** [comment]
I believe that it should not be required, but I'll leave it up to @tjfontaine

**Comment by markjdb** [comment]
(The reason it's used on FreeBSD, at least, is to enable using dtrace -h or dtrace -G without the DTrace kernel modules loaded. Otherwise dtrace processes the /usr/lib/dtrace scripts unnecessarily, and errors out when it tries to talk to the DTrace kernel code.)

**Comment by justinmk** [comment]
@tjfontaine If you have a moment, could you comment on this?

**Comment by geertj** [comment]
_bump_ build still broken on Fedora 20

**Comment by geertj** [comment]
Since @markjdb mentions that the -xnolibs option is used on FreeBSD, I suggest disabling it only on Linux:

```
diff --git a/Makefile.am b/Makefile.am
index 07d628d..7d7ae82 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -328,9 +328,14 @@ pkgconfig_DATA = @PACKAGE_NAME@.pc
 endif

 if HAVE_DTRACE
+if LINUX
+include/uv-dtrace.h: src/unix/uv-dtrace.d
+       $(AM_V_GEN)$(DTRACE) $(DTRACEFLAGS) -h -s $< -o $(top_srcdir)/$@
+else
 include/uv-dtrace.h: src/unix/uv-dtrace.d
        $(AM_V_GEN)$(DTRACE) $(DTRACEFLAGS) -h -xnolibs -s $< -o $(top_srcdir)/$@
 endif
+endif

 if DTRACE_NEEDS_OBJECTS
 SUFFIXES = .d
```

**Comment by ghost** [comment]
There's no such thing as "dtrace >= 2.5" (sic).  The -x options to dtrace(1m) are documented and form a committed interface.  If something out there purports to be DTrace but has removed them, it is not DTrace and we should not be invoking it as if it were.  This does not seem to be a bug in libuv.

**Comment by nalimilan** [comment]
I have no idea what's the real DTrace interface, but Linux distributions use SystemTap. See for Fedora:
https://admin.fedoraproject.org/pkgdb/package/systemtap/
http://linuxmanpages.net/manpages/fedora20/man1/dtrace.1.html

**Comment by ghost** [comment]
The dtrace(1m) interface is defined by the DTrace implementation.  The canonical DTrace implementation comes from the repository of record, illumos.  See https://github.com/illumos/illumos-gate/.  If GNU/Linux distributions are shipping a dtrace(1m) program that does not conform to the implementation in the DTrace repository of record, they are doing their customers a disservice.  They need to either conform to the official interface or stop calling the program 'dtrace'.

**Comment by markjdb** [comment]
Indeed. It's kind of astonishing that Fedora ships a dtrace(1) that claims to be "DTrace compatible" and only implements a subset of the dtrace(1) parameters.

**Comment by geertj** [comment]
@markjdb @wesolows why is it a problem that dtrace on Solaris behaves differently than it does on Linux? It's not like "ls" or "cp" have exactly the same flags. I admit, using the same name is somewhat confusing. But in the end it's the choice of the SystemTap developers how they name things.  And since it's open source you can always contribute patches, I'm sure the ST developers would appreciate that :)

I also don't see the benefit for the libuv project of not compiling on any recent Linux distribution. Therefore I suggest applying this patch.

**Comment by nalimilan** [comment]
Right. This is really not the place to debate whether SystemTap developers and Linux distributions were right to ship a semi-compatible command called `dtrace` (I guess it makes life easier for people using only basic options. ). Feel free to complain to them if you want.

But I don't see any problem for libuv to use a different command on the Linux platform. Where it called `stap` instead of `dtrace`, the problem would be the same (and indeed a `stap` command exists too).

**Comment by markjdb** [comment]
Because dtrace(1) comes from Solaris. There are many open-source projects which support DTrace and SystemTap probes; for SystemTap, it makes sense to leverage the use of dtrace(1) that is already present in many build scripts. But the dtrace(1) reimplementation that was apparently done for SystemTap is incomplete, so it's not a compatible implementation.

Now there are patches being submitted which break dtrace(1) usage on other operating systems, all because ST has an incomplete implementation. If they're going to leverage existing support for dtrace(1), it would be nice if it were done properly. :)

It seems reasonable to me to predicate the use of -xnolibs on the OS. But FWIW there's also a DTrace on Linux project, which I believe provides a compatible dtrace(1).

**Comment by nalimilan** [comment]
I'm not sure why they removed the `-x` option, I can't find an explanation anywhere. Before that the compatibility was apparently better.

If you want to test whether `dtrace` is actually systemtap, this works here:

```
$ dtrace --help | grep -c systemtap
1
```

**Comment by ghost** [comment]
Well, you also have the fact that _real_ DTrace on GNU/Linux (which presumably Oracle ships, although I don't use their product so I can't be sure) won't have this problem.  So just checking the equivalent of uname isn't right either.  This is an OS bug, not a libuv bug.  Please file this with Fedora instead.

**Comment by nalimilan** [comment]
@wesolows I just gave an easy way to check whether `dtrace` is actually SystemTap or the real DTrace. As regards Fedora, the bug is already fixed via an ugly call to `sed` in the .spec file. But all distributions are going to suffer from the same problem, and Fedora's policy is to try improving upstream so that everybody benefits, including people building libuv from source (potentially to contribute patches...).

Anyway, what would you want distributions to do? If they stop shipping `dtrace`, then libuv will have to include new code to work without it, or with SystemTap. How is that better?

**Comment by ghost** [comment]
It's better, at minimum, for 2 reasons:
1. It ends efforts to 'embrace and extinguish' DTrace by pretending that SystemTap is anything remotely comparable to DTrace.  In the process, it makes it easy for users to know what their system actually contains and what its true capabilities are.
2. It enables consumers like libuv to handle various different tracing frameworks that may be present without having to resort to hacks like the one you noted in order to tell which one is "really" available.  Note that it's already necessary to cope with the complete absence of a dtrace(1m) program on the system, so that piece should be a no-op.

In short, the name of a program is an interface.  It defines how a reasonable consumer expects the program to work and what capabilities they reasonably expect it to offer.  GNU's co-option of Unix was a travesty, not a precedent.  Please don't encourage others to continue the practice.  By doing the right thing here, we make our own lives much easier, our own code clearer and more maintainable, and help ourselves and others establish the correct way to factor feature detection at build time.  This is really a no-brainer and I'm somewhere between discouraged and disgusted that there's any debate at all here.

**Comment by justinmk** [comment]
> It ends efforts to 'embrace and extinguish' DTrace

It doesn't, but rather forces downstream projects like https://github.com/neovim/neovim to add their own hacks in order to provide a less-frustrating experience to users and developers.

Also, it's ironic that GNU people could make the converse argument about clang vs gcc.

**Comment by geertj** [comment]
This argument is going nowhere. Options:
1. Add a 2 line patch to make libuv compile on recent Linux versions.
2. Go on a quest to ask the systemtap developers to rename their binary based on a phylosophical argument that whatever Sun (at the time) implemented for dtrace(1m) should be the canon of how any other program claiming the same name should behave.

My vote is for 1. I'm asking that one of the core devs @saghul @bnoordhuis @piscisaureus and @indutny make a decision here.

**Comment by ghost** [comment]
I've already explained (as has @markjdb) why the "2 line patch" is wrong.  We don't accept patches that are wrong.

**Comment by geertj** [comment]
> I've already explained (as has @markjdb) why the "2 line patch" is wrong. We don't accept patches that are wrong.

@wesolows no you have not done that. You have stated your opinion, and myself and others on the thread do not agree with your opinion. Also I just checked out the commit history and I find 0 commits on your name. So I find it strange that you speak for the project like that.

**Comment by ghost** [comment]
Let me be clearer, then:

$(uname) == "Linux" =/=> "dtrace" lacks support for -x nolibs

Therefore a change based on the inverse assertion is incorrect.  A specific counterexample has been provided in the form of Oracle® Linux®; others are likely to exist (two likely candidates are FreeBSD's Linux emulation and illumos-joyent's LX-branded Zone environment).  Changeset integration count is not a substitute for a correct understanding of the problem and is an interesting combination of ad hominem and appeal to authority, fallacious in either case.  At minimum, those insisting on your preferred course of action need to demonstrate that the proposed change is beneficial or harmless in environments where uname(1) reports "Linux" but dtrace(1m) supports and likely requires -x nolibs in this context.  I have just reread the comments but did not note any such evidence.

**Comment by saghul** [comment]
Folks, relax. I'll have a look as soon ad I can, but I'm no expert in DTrace, so there is that.

Now, while we support dtrace, it just adds 2 pretty silly probes, so compiling without support for it (there is a switch for configure) should do until we find a solution.

**Comment by bnoordhuis** [comment]
I say we remove the dtrace probes or make `--without-dtrace` the default.  We only have two probes and they're not interesting ones either.

Anyone here ever managed to track down a bug or performance issue using the libuv probes?  Me neither.

EDIT: Great minds, @saghul.

**Comment by geertj** [comment]
I think @bnoordhuis gives one possible good suggestion. Just disable DTrace by default. Also @nalimilan makes a good suggestion. Check whether there is a "real" dtrace or a systemtap one:

```
$ dtrace --help | grep -c systemtap
1
```

Based on the above, HAVE_SYSTEMTAP could be set, and then the code can become

```
if HAVE_SYSTEMTAP
include/uv-dtrace.h: src/unix/uv-dtrace.d
       $(AM_V_GEN)$(DTRACE) $(DTRACEFLAGS) -h -s $< -o $(top_srcdir)/$@
endif
if HAVE_DTRACE
include/uv-dtrace.h: src/unix/uv-dtrace.d
       $(AM_V_GEN)$(DTRACE) $(DTRACEFLAGS) -h -xnolibs -s $< -o $(top_srcdir)/$@
endif
```

This could be fragile if the output of dtrace --help changes on SystemTap

**Comment by nalimilan** [comment]
Another option is to make `--without-dtrace` the default, and add a `--with-systemtap` option which would set `HAVE_SYSTEMTAP` and thus enable the code by @geertj in the previous comment.

**Comment by saghul** [comment]
+1 on removing the dtrace probes.

On Tue, Nov 4, 2014 at 8:25 PM, Ben Noordhuis notifications@github.com
wrote:

> I say we remove the dtrace probes or make --without-dtrace the default.
> We only have two probes and they're not interesting ones either.
> 
> Anyone here ever managed to track down a bug or performance issue using
> the libuv probes? Me neither.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/joyent/libuv/issues/1478#issuecomment-61745679.

## 

/Saúl
bettercallsaghul.com

**Comment by bnoordhuis** [comment]
libuv/libuv#3

**Comment by saghul** [comment]
DTrace probes are gone as of cb51400, closing.

**Comment by nalimilan** [comment]
Thanks!
